### PR TITLE
Add result support to angelscript function call

### PR DESCRIPTION
**Pull request by pat2nav**
Hi Urho team

It's very usefull to call an angelscript function and get the return of the function

Per exemple in my game, I call this AS function to get player node

```
URHO::Variant Result;
ScriptFile->Execute("Node @ CreatePlayer(Scene @ scene)", Parameters, URHO::VariantType::VAR_PTR, Result);
URHO::Node * Node = (URHO::Node*) Result.GetPtr();
```

So I propose to add 3 new functions to scriptfile with 2 new parameters

**VariantType resultType**: to specify the type of result expected
**Variant & result**: contains the result of function

### CODE REVIEW

File: Source/Urho3D/AngelScript/ScriptFile.cpp [hunk 1]
```diff
@@ -321,6 +335,78 @@ bool ScriptFile::Execute(asIScriptFunction* function, const VariantVector& param
 
     scriptSystem->IncScriptNestingLevel();
     bool success = context->Execute() >= 0;
```
Comment (weitjong) [inline Source/Urho3D/AngelScript/ScriptFile.cpp]: Based on the Angelscript documentation, the script function returns value is only valid when the execution is actually finished (asEXECUTION_FINISHED).

```suggestion
     bool success = context->Execute() == asEXECUTION_FINISHED;
```

File: Source/Urho3D/AngelScript/ScriptFile.h [hunk 2]
```diff
@@ -81,14 +81,24 @@ class URHO3D_API ScriptFile : public Resource, public ScriptEventListener
 
     /// Query for a function by declaration and execute if found.
     bool Execute(const String& declaration, const VariantVector& parameters = Variant::emptyVariantVector, bool unprepare = true);
+    /// Query for a function by declaration and execute and return result if found.
```
Comment (weitjong) [inline Source/Urho3D/AngelScript/ScriptFile.h]: Is it possible to inspect AS context to find out what is the type of the return value? Passing the return type on each call here seems not that user-friendly.

Comment (pat2nav) [reply Source/Urho3D/AngelScript/ScriptFile.h]: Ok It seems possible to get return value using asIScriptFunction::GetReturnTypeId()

File: Source/Urho3D/AngelScript/ScriptFile.cpp (hunk 1 above)
Comment (pat2nav) [reply Source/Urho3D/AngelScript/ScriptFile.cpp]: Ok fixed

**Comment by weitjong** [comment]
Just in case you forgot it, you need to push the changes for us to see them.

**Comment by pat2nav** [comment]
Finally I propose to let the bool return for error checking and get script return value as variant pointer
with automatic type detection

`bool Execute(const String& declaration, const VariantVector& parameters = Variant::emptyVariantVector, Variant* functionReturn=nullptr, bool unprepare = true);`

**Comment by pat2nav** [comment]
> 
> 
> Just in case you forgot it, you need to push the changes for us to see them.

Sorry I'm learning how to make github pull request

**Comment by ArnisLielturks** [comment]
Cool idea! This is something that would make my life easier. What about support for LUA to do the same?

**Comment by pat2nav** [comment]
> 
> 
> Cool idea! This is something that would make my life easier. What about support for LUA to do the same?

Sorry I am not familiar with Lua scripting in urho

**Comment by weitjong** [comment]
Since this is not about Urho API binding, I don't think there is a requirement that we need to have feature parity between them.

**Comment by weitjong** [comment]
Thanks again!

**Pull request closed without merging**